modules/system/boot/initmd.nix

```diff
@@ -0,0 +1,213 @@
+    cStringLit = val: "\"${builtins.replaceStrings ["\\" "\"" "\n"] ["\\\\" "\\\"" "\\n"] val}\"";
+    mkFilesystemMountC' = fs: let
+        entries = [
+            {
+                name = "fstype";
+                valueLit = fs.fsType;
+            }
+            {
+                name = "from";
+                valueLit = fs.device;
+            }
+            {
+                name = "fspath";
+                valueLit = fs.target;
+            }
+            {
+                name = "errmsg";
+                valueVar = "errmsg";
+                valueLen = "sizeof(errmsg)";
+            }
+        ] ++ builtins.map (option: let
+            split = lib.strings.splitString "=" option;
+            name = builtins.elemAt split 0;
+            values = lib.lists.sublist 1 ((builtins.length split) - 1);
+            valueLit = lib.strings.concatStringsSep "=" values;
+        in { inherit name valueLit; }) fs.options;
+        toEntryC = entry: "{ .iov_base = (void*)${cStringLit entry.name}, .iov_len = ${mountStrlen entry.name} }, { .iov_base = (void*)${if entry ? valueLit then cStringLit entry.valueLit else entry.valueVar}, .iov_len = ${if entry ? valueLit then mountStrlen entry.valueLit else entry.valueLen} }";
+        entriesList = lib.strings.concatStringsSep ", " (builtins.map toEntryC entries);
+        count = builtins.toString ((builtins.length entries)*2);
+    in ''
+        ${concatMapStrings (dep: ''
+            mkdir(${cStringLit dep}, 0755);
+        '') fs.depends}
+        mkdir(${cStringLit fs.target}, 0755);
+        if (nmount((struct iovec[${count}]){${entriesList}}, ${count}, 0) < 0) {
+            printf("init0: nmount: %s:%s onto %s: %s (%s)\n", ${cStringLit fs.fsType}, ${cStringLit fs.device}, ${cStringLit fs.target}, errmsg, strerror(errno));
+            return 21;
+        }
+        printf("init0: %s:%s onto %s: success\n", ${cStringLit fs.fsType}, ${cStringLit fs.device}, ${cStringLit fs.target});
+    '';
+    mkFilesystemMountC = fs: let
+        split = lib.strings.splitString ":" fs.device;
+        bottomLayerIdx = (builtins.length split) - 1;
+        bottomLayer = builtins.elemAt split bottomLayerIdx;
+        unionLayers = lib.lists.reverseList (lib.lists.sublist 0 bottomLayerIdx split);
+
+        bindMount' = mkFilesystemMountC' { inherit (fs) target depends; fsType = "nullfs"; device = bottomLayer; options = []; };
+        bindMount = lib.optionalString (bottomLayer != fs.target) bindMount';
+        unionMount = (lib.concatMapStrings (layer: mkFilesystemMountC' { inherit (fs) target; fsType = "unionfs"; device = layer; depends = []; options = []; }) unionLayers);
+    in if fs.fsType == "unionfs" then bindMount + unionMount else mkFilesystemMountC' fs;
+    mkFilesystemMountptC = target: mkFilesystemMountC (config.fileSystems.${target} // { inherit target; });
+    critMountsC = lib.concatMapStrings mkFilesystemMountptC sortedCritMountpointsNoRoot;
+
+    init0_src = ''
```

Comment:
I think it might be possible to use `init -r` for the reroot instead of this, but if it works it works and I'm not in the mood to rewrite it